pkg/system/autoscaler.go

```diff
@@ -0,0 +1,269 @@
+package system
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	nbv1 "github.com/noobaa/noobaa-operator/v5/pkg/apis/noobaa/v1alpha1"
+	"github.com/noobaa/noobaa-operator/v5/pkg/bundle"
+	"github.com/noobaa/noobaa-operator/v5/pkg/options"
+	"github.com/noobaa/noobaa-operator/v5/pkg/util"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/sirupsen/logrus"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	kedaHpaNoobaaEndpoint = "keda-hpa-noobaa-endpoint"
+	noobaaEndpoint        = "noobaa-endpoint"
+	hpav2NoobaaEndpoint   = "hpav2-noobaa-endpoint"
+)
+
+var autoscalerNameMap = map[nbv1.AutoscalerTypes]string{
+	nbv1.AutoscalerTypeHPAV1: noobaaEndpoint,
+	nbv1.AutoscalerTypeKeda:  kedaHpaNoobaaEndpoint,
+}
+
+func (r *Reconciler) reconcileAutoscaler() error {
+	log := r.Logger.WithField("func", "reconcileAutoscaler")
+	var err error
+	autoscalerType := nbv1.AutoscalerTypes(options.DefaultAutoscalerType)
+	if r.NooBaa.Spec.Autoscaler.AutoscalerType != "" {
+		autoscalerType = r.NooBaa.Spec.Autoscaler.AutoscalerType
+	}
+	if err = r.cleanPreviousAutoscalingResources(log, autoscalerType); err != nil {
+		return err
+	}
+	prometheusNamespace := options.PrometheusNamespace
+	if r.NooBaa.Spec.Autoscaler.PrometheusNamespace != "" {
+		prometheusNamespace = r.NooBaa.Spec.Autoscaler.PrometheusNamespace
+	}
+	switch autoscalerType {
+	case nbv1.AutoscalerTypeKeda:
+		prometheus, err := getPrometheus(log, prometheusNamespace)
+		if err != nil {
+			return err
+		}
+		if err = r.autoscaleKeda(prometheus); err != nil {
+			return err
+		}
+	case nbv1.AutoscalerTypeHPAV2:
+		//TODO
+		log.Infof("HPAV2 implementation pending")
+	case nbv1.AutoscalerTypeHPAV1:
+		if err = r.ReconcileObject(r.HPAEndpoint, r.SetDesiredHPAEndpoint); err != nil {
+			return err
+		}
+	default:
+		log.Errorf("❌ Autoscaler type is not defined for noobaa")
+		return errors.New("Autoscaler type is not defined")
+	}
+	return nil
+}
+
+func (r *Reconciler) cleanPreviousAutoscalingResources(log *logrus.Entry, autoscalerType nbv1.AutoscalerTypes) error {
```

Comment:
The cleanup code can still be much cleaner in my opinion. maybe I'm missing something, but I don't see the benefit of listing all types of scalers on every reconciliation and aggregating it into a single list (e.g.: if `autoscalerType` is hpav2, why do we need to list it for cleanup?). we should only list and clean for each type if it's not the current `autoscalerType`.
Also, I think we should not continue reconciling the autoscaler before other types are cleaned, so we won't have 2 different autoscalers configuring the endpoints deployment. so if any other autoscalers exist, then the cleanup function should return a temp error.  maybe also rename the function to something that better states that we are blocked for the cleanup, something like `ensurePreviousAutoscalerCleanup`

I suggest something like that:

```golang
func (r *Reconciler) ensurePreviousAutoscalerCleanup(autoscalerType nbv1.AutoscalerTypes) error {
   if autoscalerType != nbv1.AutoscalerTypeHPAV1 {
      if err := r.ensureHPAV1Cleanup(); err != nil {
         return err
      }
   }
   if autoscalerType != nbv1.AutoscalerTypeHPAV2 {
      if err := r.ensureHPAV2Cleanup(); err != nil {
         return err
      }
   }
   if autoscalerType != nbv1.AutoscalerTypeHPAV1 {
      if err := r.ensureKEDACleanup(); err != nil {
         return err
      }
   }

   return nil
}


func (r *Reconciler) ensureHPAV1Cleanup() error {
   noobaaHpaSelector, _ := labels.Parse("app=noobaa")
   //get autoscaler only for HPAV1 based on the label
   autoscalersv1 := &autoscalingv1.HorizontalPodAutoscalerList{}
   util.KubeList(autoscalersv1, &client.ListOptions{Namespace: r.Request.Namespace, LabelSelector: noobaaHpaSelector})
   // ... delete all autoscaresv1. I don't see a reason to look for a specific name, since you already list based on the label.

   
   if len(autoscalersv1) > 0 {
      //As long as we still have HPAV1 running we don't want to continue. return error
      return fmt.Errorf("HPAV1 autoscaler still not deleted ")
   }

}


// same for HPAV2\KEDA
```


notice that I also removed the logger parameter. we already have it on the Reconciler struct